src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,3106 @@
+ * {
+ *	int i;
+ *
+ *	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+ *		if (fd_dup2_list[i].fd_src == fd)	{
+ *			return 1;
+ *		}
+ *	}
+ *	return 0;
+ * }
+ */
+static int
+query_fd_forward_dest(int fd_src)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+		if (fd_src == fd_dup2_list[i].fd_src)
+			return fd_dup2_list[i].fd_dest;
+	}
+	return -1;
+}
+
+static int
+allocate_a_fd_from_kernel(void)
+{
+	struct timespec	times_loc;
+	char		file_name[128];
+
+	clock_gettime(CLOCK_REALTIME, &times_loc);
+	snprintf(file_name, sizeof(file_name) - 1, "dummy_%ld_%ld", times_loc.tv_sec,
+		 times_loc.tv_nsec);
+	return memfd_create("dummy", 0);
+}
+
+static void
+close_all_duped_fd(void)
+{
+	int i;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+		if (fd_dup2_list[i].fd_src >= 0)	{
+			fd_dup2_list[i].fd_src = -1;
+			new_close_common(real_close_libc, fd_dup2_list[i].fd_dest);
+		}
+	}
+}
+
+/**
+ *static int
+ * Query_Fd_Forward_Src(int fd_dest, int *Closed)
+ * {
+ *	int i;
+ *
+ *	for (i = 0; i < MAX_FD_DUP2ED; i++)	{
+ *		if (fd_dest == fd_dup2_list[i].fd_dest)	{
+ *			return fd_dup2_list[i].fd_src;
+ *		}
+ *	}
+ *	return -1;
+ * }
+ */
+static int
+open_common(int (*real_open)(const char *pathname, int oflags, ...), const char *szCallerName,
+	    const char *pathname, int oflags, ...)
+{
+	unsigned int mode = 0664;
+	int two_args = 1, rc, is_target_path, idx_fd, idx_dirfd;
+	char item_name[MAX_FILE_NAME_LEN];
+	char parent_dir[MAX_FILE_NAME_LEN];
+	char full_path[MAX_FILE_NAME_LEN];
+	dfs_obj_t *file_obj;
+	dfs_obj_t *parent;
+	mode_t mode_query = 0;
+
+	if (oflags & O_CREAT)   {
+		va_list arg;
+
+		va_start(arg, oflags);
+		mode = va_arg(arg, unsigned int);
+		mode = mode & mode_not_umask;
+		va_end(arg);
+		two_args = 0;
+	}
+
+	parse_path(pathname, &is_target_path, &parent, item_name, parent_dir, full_path);
+	if (is_target_path) {
+		/* file/dir should be handled by our FS */
+		if (oflags & O_CREAT) {
+			rc = dfs_open(dfs, parent, item_name, mode | S_IFREG, oflags, 0, 0,
+				      NULL, &file_obj);
+			mode_query = S_IFREG;
+		} else if (!parent && (strncmp(item_name, "/", 2) == 0)) {
+			rc = dfs_lookup(dfs, "/", oflags, &file_obj, &mode_query, NULL);
+		} else {
+			rc = dfs_lookup_rel(dfs, parent, item_name, oflags, &file_obj, &mode_query,
+					    NULL);
+		}
+
+		if (rc) {
+			printf("open_common> Error: Fail to dfs_open/dfs_lookup_rel %s rc = %d\n",
+			       pathname, rc);
+			errno = rc;
+			return (-1);
+		}
+		if (S_ISDIR(mode_query)) {
+			idx_dirfd = find_next_available_dirfd();
+			assert(idx_dirfd >= 0);
+
+			dir_list[idx_dirfd].fd = idx_dirfd + FD_DIR_BASE;
+			dir_list[idx_dirfd].offset = 0;
+			dir_list[idx_dirfd].fd_dup_pre = -1;
+			dir_list[idx_dirfd].fd_dup_next = -1;
+			dir_list[idx_dirfd].ref_count = 1;
+			dir_list[idx_dirfd].dir_obj = file_obj;
+			dir_list[idx_dirfd].num_ents = 0;
+			memset(&dir_list[idx_dirfd].anchor, 0, sizeof(daos_anchor_t));
+			sprintf(dir_list[idx_dirfd].path, "%s%s", fs_root, full_path);
```

Comment:
This feels like another case for asprintf() unless you have a check for max path length.